### PR TITLE
config.py: don't rely on section names

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
In `config.py`, stop relying on section names for `config.py full`, `config.py realfull`, etc. This way we can move configuration settings around as described in https://github.com/Mbed-TLS/mbedtls/pull/9236 without having to worry about messing up the tooling.

Specifically:

* `full` and friends no longer rely on the section to determine which settings are booleans that should be activated.
* `realfull` is now really full. The goal is to document everything, and it turns out that uncommenting everything does work.

## Testing

The script `tests/scripts/test_config_script.py` runs some tests on `config.py` and places the output in the given directory. Run it on the version of `config.py` from two commits to compare its effects.

```
$ git checkout --recurse-submodules upstream/pull/9604~4
$ tests/scripts/test_config_script.py -d 0
$ git checkout --recurse-submodules upstream/pull/9604~3
$ tests/scripts/test_config_script.py -d 1
$ git checkout --recurse-submodules upstream/pull/9604~2
$ tests/scripts/test_config_script.py -d 2
$ git checkout --recurse-submodules upstream/pull/9604~1
$ tests/scripts/test_config_script.py -d 3
$ git checkout --recurse-submodules upstream/pull/9604
$ tests/scripts/test_config_script.py -d 4
$ diff -ru 0 1
$ diff -ru 1 2
$ diff -ru 2 3 | less
$ diff -ru 3 4
```

2→3 has an expected difference in the behavior of `realfull`.

## PR checklist

- [x] **changelog** not required because: refactoring only
- [x] **development PR** provided # | not required because: 
- [x] **framework PR** https://github.com/Mbed-TLS/mbedtls-framework/pull/49
- [x] **3.6 PR** https://github.com/Mbed-TLS/mbedtls/pull/9637
- [x] **2.28 PR** not required because: we won't change `config.py` in 2.28
- **tests**  manual (see above)

### CODE REVIEW

File: scripts/config.py
Comment (mpg) [inline scripts/config.py:206]: Shouldn't we use the `is_boolean_setting()` function introduced a few commits earlier? We're doing the same here (`if not value` then `if name.startswith('PSA_WANT_')`.

File: include/mbedtls/ssl.h
Comment (mpg) [inline include/mbedtls/ssl.h:442]: The settings above have no doxygen documentation in `ssl.h`, but have it in `mbedtls_config.h` instead. So I'm not sure I understand the commit message: is it a new rule, or an existing rule that I didn't know and that wasn't consistently followed so far?

Also, why move the documentation specifically for this option but not the others? Overall I'm not sure I understand how this commit relates to the main goal of the PR.

**Review by mpg — COMMENTED** [review]
I'm happy with the general approach, but there's a few things I need clarified before I can approve.

File: include/mbedtls/ssl.h
Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:442]: It's an existing rule, but not one that we've followed consistently, or ever bothered to write down. Some settings, including all defined-or-not settings, are documented in `mbedtls_config.h` (formerly `config.h`), and `config.py realfull` (formerly `config.pl realfull`) activates all of them so that they appear in the official documentation. Other settings, which are macros with a value, are documented in a topical header file, and wrapped with `#if !defined` in that header file. We don't activate those settings in `realfull` because otherwise Doxygen would see the documentation twice and complain.

Before this PR, documented-in-config.h was based on sections. After this PR, it's mostly based on has-value vs boolean, which I think is easier to understand.

We should make the rule clearer, but that's a preexisting issue outside the scope of this PR, which is about facilitating the implementation of the specification in https://github.com/Mbed-TLS/mbedtls/pull/9236.

As for why we have this rule, I call on the author of 2134d25070a635db3770ec607487380954879499..

Comment (mpg) [reply include/mbedtls/ssl.h:442]: I'm sorry, I don't get how the rule you mention would follow from https://github.com/Mbed-TLS/mbedtls/commit/2134d25070a635db3770ec607487380954879499 - quite the contrary, the commit message suggests `config.h` had things like:
```
/** Some doxygen doc */
//#define MBEDTLS_FOO
```
which doxygen complains about and we're trying to pacify it by uncommenting the `#define`, not by moving the doxygen comment elsewhere.

I agree that as long as doxygen documentation is generated using `realfull`, macros that are commented out by default in config.h need to have their doxygen doc either in `mbedtls_config.h` or in a topical header (eg `ssl.h`) but not both. I don't see how it logically follows that it must be the 2nd one. I wasn't aware of any rule that it must be the 2nd one, and since you say it's unwritten, I'd say it's hard to know if "we" had such a rule.

Ah, taking a closer look, `realfull` is not doing exactly what I thought it was doing: it's uncommenting everything in most sections and leaving things commented out the "Module configuration options" section, and its docstring says that's because those should be documented in their respective headers - it turns out we have some written rule after all. I'll note that was written by the author of ba4162a52631c0d3ec342a91a0d8924410b4463a - not me. See also the comment maintainers near the start of the "Module configuration options" section, which states something similar but also suggests duplicating the doxygen comment in `config.h`.

Honestly that seems quite complicated. In practice several options don't respect that rule (most SSL options have no doxygen in `ssl.h`) and those that respect the rule have their doxygen duplicated in `mbedtls_config.h` and their topical header, with the the usual issues that come with duplication. I can't see a good reason for that. IMO it would be much simpler for everything to be documented in a single place, and for that single place to be `mbedtls_config.h` - no distinction between sections, experience shows we're not good at remembering those.

Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:442]: In ba4162a52631c0d3ec342a91a0d8924410b4463a, I was describing existing behavior, not giving its rationale. I'm now thinking that the current behavior of `realfull` depending on sections is a historical accident and not something we needed for a technical purpose.

Comment (mpg) [reply include/mbedtls/ssl.h:442]: Yes, I think things evolved in a messy way and a various points one of us tried fixing/improving/documenting things based on the current state at that time, but apparently before today we never took the time to really think about the whole picture :(

**Comment by gilles-peskine-arm** [comment]
I have rebased twice:

1. Once from https://github.com/gilles-peskine-arm/mbedtls/tree/config-full-booleans-only-1 to https://github.com/gilles-peskine-arm/mbedtls/tree/config-full-booleans-only-2 to cancel the changes related to preserving the not-actually-full behavior of `realfull`, following the discussion with Manuel here and in Slack.
2. Once because now that #9470 is merged, part of the changes are now in the framework. (With an extra force-push that just changes the commit date, because I forgot to push the framework branch before triggering the CI.)

**Review by mpg — COMMENTED** [review]
I'm happy with the rebases but I don't think my previous feedback has been addressed.

File: scripts/config.py
Comment (gilles-peskine-arm) [reply scripts/config.py]: I believe this comment is moot, given the complete redesign of `realfull`. If this is still relevant, please clarify.

File: scripts/config.py
Comment (gilles-peskine-arm) [reply scripts/config.py:41]: I believe this comment is moot, given the complete redesign of `realfull`. If this is still relevant, please clarify.

File: scripts/config.py
Comment (gilles-peskine-arm) [reply scripts/config.py]: This piece of code no longer exists after the complete redesign of `realfull`.

**Review by mpg — DISMISSED** [review]
LGTM!

I had marked some of my comments as resolved by the realfull rework, but I for some reason I failed to notice that was also the case for the others, sorry.

**Review by ronald-cron-arm — DISMISSED** [review]
LGTM

**Comment by mpg** [comment]
I've merged the framework PR, please update to point to the merge commit.

**Review by mpg — APPROVED** [review]
LGTM

**Review by ronald-cron-arm — APPROVED** [review]
LGTM

**Pull request merged**
